### PR TITLE
perf: optimize onchain plonk verifier

### PR DESCRIPTION
**Pull request by mrain**
Closes #<ISSUE_NUMBER>

### This PR:

Optimize the gas cost for PlonkVerifier library, saving: 17,667 gas (−4.6 %) on `test_verify_succeeds`.
Baseline (PlonkVerifierV3): 383,601 gas.  Optimised (PlonkVerifierV4): 365,934 gas.

### CODE REVIEW

File: contracts/src/libraries/PlonkVerifierV4.sol [hunk 1]
```diff
@@ -0,0 +1,1017 @@
...
+
+        // Validate publicInput scalars inline (saves 5 function-call JUMPs)
+        assembly {
+            let R := 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
+            for { let i := 0 } lt(i, 5) { i := add(i, 1) } {
+                if iszero(lt(mload(add(publicInput, mul(i, 0x20))), R)) {
+                    mstore(0x00, 0x05b05ccc00000000000000000000000000000000000000000000000000000000)
+                    revert(0x00, 0x04)
+                }
+            }
```
[the remark this answers is not in the document]
Comment (mrain) [reply contracts/src/libraries/PlonkVerifierV4.sol:150]: I'm dumb here. @alysiahuggins @sveitser Claude told me that `publicInput` is a fixed-size array, so there'll be no array length at the prefix. Which one is correct

**Comment by mrain** [comment]
> @mrain this is quite a lot of difficult to review assembly code. I don't think anyone can realistically review this for correctness.

This PR is mostly vibed, I did my review by comparing it to the existing PlonkVerifierV3.

> 
> 1. Let's calculate how much this would save us per year, in USD at current gas prices and LC update interval. I think it may not be worth it.
> 2. If we still want it we need to set up a test system that compares input and output of existing and optimized code version that covers all edge cases.
> 
> It seems currently the "expensive" transactions to the LC cost about USD1 so for 2 txns per day this is less than USD 1k per year. 5% of that is USD 50 savings per year.

Yes I agree with you here, I'll put it in draft now so we can pick it up once it's needed.

**Comment by sveitser** [comment]
Another problem with our assembly code is that it makes it impossible to run coverage analysis on our smart contract code. So generally I think it's better to not use assembly unless it's very small core functions that get called a lot. The reason it breaks coverage is that unless the assembly blocks are annotated as memory-safe the compiler has less options for code generation. Figuring out if assembly blocks are memory safe I find really hard.
